MON-4361,MON-4380: Optional Monitoring Capability

rexagod opened the pull request:
Metric rules and metrics exporters have not been opted-in to keep the telemetry rules functioning.

<!--
    Don't forget about CHANGELOG if this affects the end user!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Monitoring <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR
    <Component> Component affected by your changes such as deps bump, alerts changes and any user facing changes.

    Example:
    - [#741](https://github.com/openshift/cluster-monitoring-operator/pull/741) Bump thanos components to v0.11.0 release
-->

* [x] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

simonpasquier reviewed (commented):
It might be good to wait for https://github.com/openshift/cluster-monitoring-operator/pull/2649 since it's migrating all the dashboards to static assets.

On the jsonnet implementation side, I wonder if it wouldn't easier to read/maintain if we inject the annotation into each component that needs it.
E.g. here for components for which all resources are OptionalMonitoring

https://github.com/openshift/cluster-monitoring-operator/blob/ea9a533f05318f51114d2e98e5172f1dcd769eb4/jsonnet/main.jsonnet#L520-L539

Or at the level of the jsonnet component file in case it's per resource.

simonpasquier commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@
   - `KubePdbNotEnoughHealthyPods`
   - `KubeNodePressure`
   - `KubeNodeEviction`
+- []() Allow cluster-admins to opt-into optional monitoring using the `OptionalMonitoring` capability.
```
I realize that adding the annotation to the manifests under the `assets/` directory will have no direct effect since there's no logic in CMO to deploy these resources conditionally, right?

rexagod committed:
MON-4361: Annotate optional monitoring resources

Metric rules and metrics exporters have not been opted-in to keep the
telemetry rules functioning. Optional components include:
* Alertmanager
* AlertmanagerUWM
* ClusterMonitoringOperatorDeps (partially, for AM)
* MonitoringPlugin
* PromtheusOperator (partially, for AM)
* PromtheusOperatorUWM
* ThanosRuler

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod commented:
Reverted the `capability.openshift.io/name: Console` to respect [being able to support dashboards in optional monitoring](https://redhat-internal.slack.com/archives/C6A3NV5J9/p1753979917745059?thread_ts=1753965750.365569&cid=C6A3NV5J9) since we'll still be scraping all targets anyway (to not break any telemetry rules).

rexagod committed:
MON-4361,MON-4380: Opt `monitoring-plugin` into Console capability

Drop `monitoring-plugin` from optional components as its deployment
should only rely on the `Console` capability. This is in-line with its
corresponding task's behavior as well. Also refactored the `jsonnet`
code a bit.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod committed:
MON-4361,MON-4380: Add optional monitoring logic

Enabling the `OptionalMonitoring` capability translates to enabling all
optional monitoring components under CMO. Note that since capabilities
cannot be disabled once enabled, so cleanup for optional monitoring
resources is not necessary. To clarify further, there are two possible
paths at install time:
* capability is disabled -> enabled (no need to cleanup)
* capability is enabled -/> (cannot be disabled) (no need to cleanup)

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod committed:
MON-4361,MON-4380: Make CRDs non-optional

Excluding CRDs from optional monitoring as their absence can cause CMO
and PO to crash or throw errors at the very least.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod committed:
MON-4361,MON-4380: Add webhook description and refactor jsonnet/

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod committed:
MON-4361,MON-4380: Refactor optional monitoring logic

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod commented:
/remove-lifecycle stale

rexagod commented:
/remove-lifecycle stale

Will revisit this once the related telemetry-profile effort is complete.
